File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.js
let fs = require("fs");
let input = fs.readFileSync("/dev/stdin");
// let input = fs.readFileSync("./input.txt");
input = input
  .toString()
  .trim()
  .split(" ")
  .map((x) => Number(x));

const N = input[0];
const K = input[1];

dp = new Array(100001).fill(Infinity);
ans = Infinity;

function bfs(n, sec) {
  q = [[n, sec]];
  dp[n] = 0;

  while (q.length) {
    let [x, y] = q.shift();

    if (y > ans) {
      continue;
    }

    if (x === K) {
      ans = Math.min(ans, y);
    }

    if (x + 1 <= 100000) {
      if (dp[x + 1] > y + 1) {
        dp[x + 1] = y + 1;
        q.push([x + 1, y + 1]);
      }
    }

    if (x - 1 >= 0) {
      if (dp[x - 1] > y + 1) {
        dp[x - 1] = y + 1;
        q.push([x - 1, y + 1]);
      }
    }

    if (2 * x >= 0 && 2 * x <= 100000) {
      if (dp[2 * x] > y) {
        dp[2 * x] = y;
        q.push([2 * x, y]);
      }
    }
  }
}

bfs(N, 0);
console.log(ans);

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
import sys
from collections import deque
# sys.stdin = open("input.txt", 'r')
input = sys.stdin.readline

N, K = map(int, input().split())
dp = [float('inf')] * 100001
ans = float('inf')
def bfs(n, sec):
    global ans
    q = deque([(n, sec)])
    dp[n] = 0
    while q:

        x, y = q.popleft()

        if x == K:
            ans = min(ans, y)


        if x+1 <= 100000:
            if dp[x+1] > y+1:
                dp[x+1] = y+1
                q.append((x + 1, y + 1))
        if x-1 >= 0:
            if dp[x-1] > y+1:
                dp[x-1] = y+1
                q.append((x - 1, y + 1))
        if 0 <= 2 * x <= 100000:
            if dp[2*x] > y:
                dp[2*x] = y
                q.append((2 * x, y))

bfs(N, 0)
print(ans)
